File: Packages/RallyUI/Sources/RallyUI/Components/RewardRow.swift
import SwiftUI
import RallyCore

/// Reward catalog row displaying the reward image, title, point cost, and a
/// redeem call-to-action button.
///
/// Usage:
/// ```swift
/// RewardRow(reward: reward, userPoints: 1200) {
///     redeemReward(reward)
/// }
/// ```
public struct RewardRow: View {

    // MARK: - Properties

    private let reward: Reward
    private let userPoints: Int
    private let onRedeem: (() -> Void)?

    private var canAfford: Bool {
        userPoints >= reward.pointsCost
    }

    // MARK: - Init

    /// Creates a reward row.
    /// - Parameters:
    ///   - reward: The reward to display.
    ///   - userPoints: The user's current point balance, used to determine CTA state.
    ///   - onRedeem: Closure invoked when the user taps the redeem button.
    public init(
        reward: Reward,
        userPoints: Int = 0,
        onRedeem: (() -> Void)? = nil
    ) {
        self.reward = reward
        self.userPoints = userPoints
        self.onRedeem = onRedeem
    }

    // MARK: - Body

    public var body: some View {
        HStack(spacing: RallySpacing.smMd) {
            // Reward image
            rewardImage

            // Text content
            VStack(alignment: .leading, spacing: RallySpacing.xs) {
                Text(reward.title)
                    .font(RallyTypography.cardTitle)
                    .foregroundStyle(.white)
                    .lineLimit(2)

                HStack(spacing: RallySpacing.xs) {
                    categoryBadge

                    if let tier = tierBadge {
                        tier
                    }
                }

                // Points cost
                HStack(spacing: RallySpacing.xs) {
                    Image(systemName: "star.fill")
                        .font(.system(size: 11))
                        .foregroundStyle(RallyColors.orange)

                    Text(reward.pointsCost.pointsFormatted)
                        .font(RallyTypography.buttonLabel)
                        .foregroundStyle(canAfford ? RallyColors.orange : RallyColors.gray)
                }
            }

            Spacer()

            // Redeem CTA
            redeemButton
        }
        .padding(RallySpacing.smMd)
        .background(
            RoundedRectangle(cornerRadius: RadiusToken.card, style: .continuous)
                .fill(RallyColors.navyMid)
        )
        .rallyCardShadow()
        .accessibilityElement(children: .ignore)
        .accessibilityLabel(accessibilityLabelText)
        .accessibilityAddTraits(canAfford ? .isButton : [])
    }

    // MARK: - Subviews

    @ViewBuilder
    private var rewardImage: some View {
        if let imageURL = reward.imageURL {
            AsyncImage(url: imageURL) { phase in
                switch phase {
                case .success(let image):
                    image
                        .resizable()
                        .aspectRatio(contentMode: .fill)
                case .failure:
                    imagePlaceholder
                case .empty:
                    ProgressView()
                        .frame(width: 72, height: 72)
                @unknown default:
                    imagePlaceholder
                }
            }
            .frame(width: 72, height: 72)
            .clipShape(RoundedRectangle(cornerRadius: RadiusToken.small, style: .continuous))
        } else {
            imagePlaceholder
        }
    }

    private var imagePlaceholder: some View {
        RoundedRectangle(cornerRadius: RadiusToken.small, style: .continuous)
            .fill(RallyColors.navy)
            .frame(width: 72, height: 72)
            .overlay(
                Image(systemName: categoryIcon)
                    .font(.system(size: 24))
                    .foregroundStyle(RallyColors.gray.opacity(0.6))
            )
    }

    private var categoryBadge: some View {
        Text(reward.category.rawValue.capitalized)
            .font(RallyTypography.caption)
            .foregroundStyle(RallyColors.blue)
            .padding(.horizontal, RallySpacing.sm)
            .padding(.vertical, 2)
            .background(
                Capsule()
                    .fill(RallyColors.blue.opacity(0.12))
            )
    }

    @ViewBuilder
    private var tierBadge: some View? {
        if reward.minimumTier != .rookie {
            Text(reward.minimumTier.rawValue)
                .font(RallyTypography.caption)
                .foregroundStyle(reward.minimumTier.color)
                .padding(.horizontal, RallySpacing.sm)
                .padding(.vertical, 2)
                .background(
                    Capsule()
                        .fill(reward.minimumTier.color.opacity(0.12))
                )
        }
    }

    private var redeemButton: some View {
        Button {
            if canAfford {
                let generator = UIImpactFeedbackGenerator(style: .medium)
                generator.impactOccurred()
                onRedeem?()
            }
        } label: {
            Text(canAfford ? "Redeem" : "Need more")
                .font(RallyTypography.caption)
                .fontWeight(.semibold)
                .foregroundStyle(canAfford ? .white : RallyColors.gray)
                .padding(.horizontal, RallySpacing.smMd)
                .padding(.vertical, RallySpacing.sm)
                .background(
                    RoundedRectangle(cornerRadius: RadiusToken.button, style: .continuous)
                        .fill(canAfford ? RallyColors.orange : RallyColors.navy)
                )
        }
        .disabled(!canAfford)
        .accessibilityLabel(canAfford ? "Redeem \(reward.title)" : "Not enough points for \(reward.title)")
    }

    // MARK: - Helpers

    private var categoryIcon: String {
        switch reward.category {
        case .merchandise: return "tshirt.fill"
        case .concessions: return "cup.and.saucer.fill"
        case .experiences: return "ticket.fill"
        case .tickets: return "ticket.fill"
        case .digital: return "gift.fill"
        case .partner: return "storefront.fill"
        }
    }

    private var accessibilityLabelText: String {
        var label = "\(reward.title), \(reward.pointsCost) points"
        if canAfford {
            label += ", available to redeem"
        } else {
            label += ", need \(reward.pointsCost - userPoints) more points"
        }
        return label
    }
}

// MARK: - Preview

#Preview("Reward Rows") {
    let rewards: [Reward] = [
        Reward(
            id: "1",
            schoolID: "s1",
            title: "Team T-Shirt",
            description: "Official team merchandise",
            pointsCost: 500,
            category: .merchandise
        ),
        Reward(
            id: "2",
            schoolID: "s1",
            title: "VIP Sideline Experience",
            description: "Watch warm-ups from the sideline",
            pointsCost: 5000,
            category: .experiences,
            minimumTier: .mvp
        ),
        Reward(
            id: "3",
            schoolID: "s1",
            title: "Free Hot Dog Combo",
            description: "Hot dog, chips, and drink",
            pointsCost: 200,
            category: .concessions
        ),
    ]

    ScrollView {
        VStack(spacing: 12) {
            ForEach(rewards) { reward in
                RewardRow(reward: reward, userPoints: 1200)
            }
        }
        .padding()
    }
    .background(RallyColors.navy)
}

File: Android/ui/src/main/java/com/vanwagner/rally/ui/components/RewardRow.kt
package com.vanwagner.rally.ui.components

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.CardGiftcard
import androidx.compose.material.icons.filled.ConfirmationNumber
import androidx.compose.material.icons.filled.Fastfood
import androidx.compose.material.icons.filled.Redeem
import androidx.compose.material.icons.filled.ShoppingBag
import androidx.compose.material.icons.filled.Star
import androidx.compose.material.icons.filled.Store
import androidx.compose.material3.Icon
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import com.vanwagner.rally.ui.theme.RallyColors
import com.vanwagner.rally.ui.theme.RallyRadius
import com.vanwagner.rally.ui.theme.RallySpacing
import com.vanwagner.rally.ui.theme.RallyTypography
import com.vanwagner.rally.ui.theme.tierColor
import java.text.NumberFormat
import java.util.Locale

// ---------------------------------------------------------------------------
// RewardRow
// Reward catalog item: image placeholder, title, category badge, optional
// tier-lock badge, points cost, and redeem CTA.
// ---------------------------------------------------------------------------

/** Reward categories matching the iOS model. */
enum class RewardCategory(val displayName: String) {
    Merchandise("Merchandise"),
    Concessions("Concessions"),
    Experiences("Experiences"),
    Tickets("Tickets"),
    Digital("Digital"),
    Partner("Partner"),
}

/**
 * Reward list row.
 *
 * @param title       Reward display name.
 * @param pointsCost  Number of points required to redeem.
 * @param category    The [RewardCategory].
 * @param modifier    Optional [Modifier].
 * @param userPoints  The user's current point balance (default 0).
 * @param minimumTier Tier name required (e.g. "MVP"); `null` or "Rookie" hides badge.
 * @param onRedeem    Callback invoked when the redeem button is tapped.
 */
@Composable
fun RewardRow(
    title: String,
    pointsCost: Int,
    category: RewardCategory,
    modifier: Modifier = Modifier,
    userPoints: Int = 0,
    minimumTier: String? = null,
    onRedeem: (() -> Unit)? = null,
) {
    val canAfford = userPoints >= pointsCost
    val pointsFormatted = NumberFormat.getNumberInstance(Locale.US).format(pointsCost)

    val a11yLabel = buildString {
        append("$title, $pointsCost points")
        if (canAfford) append(", available to redeem")
        else append(", need ${pointsCost - userPoints} more points")
    }

    Row(
        modifier = modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(RallyRadius.card))
            .background(RallyColors.NavyMid)
            .padding(RallySpacing.smMd)
            .semantics { contentDescription = a11yLabel },
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(RallySpacing.smMd),
    ) {
        // Reward image placeholder
        Box(
            modifier = Modifier
                .size(72.dp)
                .clip(RoundedCornerShape(RallyRadius.small))
                .background(RallyColors.Navy),
            contentAlignment = Alignment.Center,
        ) {
            Icon(
                imageVector = categoryIcon(category),
                contentDescription = null,
                tint = RallyColors.Gray.copy(alpha = 0.6f),
                modifier = Modifier.size(24.dp),
            )
        }

        // Text content
        Column(
            modifier = Modifier.weight(1f),
            verticalArrangement = Arrangement.spacedBy(RallySpacing.xs),
        ) {
            Text(
                text = title,
                style = RallyTypography.cardTitle,
                color = Color.White,
                maxLines = 2,
                overflow = TextOverflow.Ellipsis,
            )

            // Category + tier badges
            Row(horizontalArrangement = Arrangement.spacedBy(RallySpacing.xs)) {
                // Category badge
                Text(
                    text = category.displayName,
                    style = RallyTypography.caption,
                    color = RallyColors.Blue,
                    modifier = Modifier
                        .background(
                            color = RallyColors.Blue.copy(alpha = 0.12f),
                            shape = RoundedCornerShape(50),
                        )
                        .padding(horizontal = RallySpacing.sm, vertical = 2.dp),
                )

                // Tier badge (if not Rookie / null)
                val showTier = minimumTier != null &&
                    minimumTier.lowercase() != "rookie"
                if (showTier) {
                    val tColor = tierColor(minimumTier!!)
                    Text(
                        text = minimumTier,
                        style = RallyTypography.caption,
                        color = tColor,
                        modifier = Modifier
                            .background(
                                color = tColor.copy(alpha = 0.12f),
                                shape = RoundedCornerShape(50),
                            )
                            .padding(horizontal = RallySpacing.sm, vertical = 2.dp),
                    )
                }
            }

            // Points cost
            Row(
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.spacedBy(RallySpacing.xs),
            ) {
                Icon(
                    imageVector = Icons.Filled.Star,
                    contentDescription = null,
                    tint = RallyColors.Orange,
                    modifier = Modifier.size(11.dp),
                )
                Text(
                    text = pointsFormatted,
                    style = RallyTypography.buttonLabel,
                    color = if (canAfford) RallyColors.Orange else RallyColors.Gray,
                )
            }
        }

        // Redeem CTA
        val ctaBackground = if (canAfford) RallyColors.Orange else RallyColors.Navy
        val ctaForeground = if (canAfford) Color.White else RallyColors.Gray
        val ctaText = if (canAfford) "Redeem" else "Need more"

        Text(
            text = ctaText,
            style = RallyTypography.caption.copy(fontWeight = FontWeight.SemiBold),
            color = ctaForeground,
            modifier = Modifier
                .clip(RoundedCornerShape(RallyRadius.button))
                .background(ctaBackground)
                .then(
                    if (canAfford && onRedeem != null) {
                        Modifier.clickable(onClick = onRedeem)
                    } else {
                        Modifier
                    }
                )
                .padding(horizontal = RallySpacing.smMd, vertical = RallySpacing.sm),
        )
    }
}

// -- Icon mapping --

private fun categoryIcon(category: RewardCategory): ImageVector = when (category) {
    RewardCategory.Merchandise -> Icons.Filled.ShoppingBag
    RewardCategory.Concessions -> Icons.Filled.Fastfood
    RewardCategory.Experiences -> Icons.Filled.ConfirmationNumber
    RewardCategory.Tickets     -> Icons.Filled.ConfirmationNumber
    RewardCategory.Digital     -> Icons.Filled.CardGiftcard
    RewardCategory.Partner     -> Icons.Filled.Store
}

// MARK: - Previews

@Preview(name = "RewardRow - Variants", showBackground = true, backgroundColor = 0xFF131B2E)
@Composable
private fun RewardRowPreview() {
    Surface(color = RallyColors.Navy) {
        Column(
            modifier = Modifier
                .padding(RallySpacing.md)
                .verticalScroll(rememberScrollState()),
            verticalArrangement = Arrangement.spacedBy(RallySpacing.smMd),
        ) {
            RewardRow(
                title = "Team T-Shirt",
                pointsCost = 500,
                category = RewardCategory.Merchandise,
                userPoints = 1200,
            )
            RewardRow(
                title = "VIP Sideline Experience",
                pointsCost = 5000,
                category = RewardCategory.Experiences,
                userPoints = 1200,
                minimumTier = "MVP",
            )
            RewardRow(
                title = "Free Hot Dog Combo",
                pointsCost = 200,
                category = RewardCategory.Concessions,
                userPoints = 1200,
            )
        }
    }
}
